### ISSUE TITLE
[BUG] Difference in output compared to non-Alpa run for Dropout

### ISSUE THREAD

**Issue by yf225** [alpa-projects/alpa#372]
I was trying to compare the output from Alpa and non-Alpa for Dropout function, and found the following:
```python
import jax
import alpa
from jax import numpy as jnp
from jax import lax

def dropout_func(rng, input):
    p = 0.5
    keep_prob = 1. - p
    mask = jax.random.bernoulli(rng, p=keep_prob, shape=input.shape)
    return lax.select(mask, input, jnp.zeros_like(input))

x = jnp.ones((8, 3))

rng = jax.random.PRNGKey(0)
jax_output = dropout_func(rng, x)

rng = jax.random.PRNGKey(0)
alpa_output = alpa.api.parallelize(dropout_func)(rng, x)

print("jax_output: ", jax_output)
print("alpa_output: ", alpa_output)
assert len(jax_output) == len(alpa_output)
for a, b in zip(jax_output, alpa_output):
    assert jnp.allclose(a, b, rtol=1e-3)
```
The `jnp.allclose` equality check fails, and printing out `jax_output` and `alpa_output` gives:
```
jax_output:
[[0. 1. 1.]
 [0. 1. 1.]
 [1. 1. 0.]
 [1. 0. 1.]
 [0. 1. 1.]
 [1. 0. 1.]
 [0. 1. 1.]
 [1. 0. 0.]]

alpa_output:
[[0. 1. 1.]
 [0. 1. 1.]
 [0. 1. 1.]
 [0. 1. 1.]
 [0. 1. 1.]
 [0. 1. 1.]
 [0. 1. 1.]
 [0. 1. 1.]]
```
It seems that all parallelism groups are following the RNG key of the first group. Maybe this is a bug?

**Comment by merrymercy** [comment]
Yes. Currently, alpa does not handle RNG key correctly.
Jax uses stateless RNG functions, so users explicitly control the RNG key.
It is not very easy to support this in Alpa, so we switched to stateful RNG functions.

We monkey patched `jax.random.uniform`, which is used by `nn.Dropout` in flax. But we didn't monkey patch `jax.random.bernouli`, so you see this bug.
https://github.com/alpa-projects/alpa/blob/9029f160634b54642e95bdd735a345e3e2361d25/alpa/monkey_patch.py#L45-L61
